src/core_plugins/tagcloud/public/tag_cloud_visualization.js

```diff
@@ -0,0 +1,121 @@
+import TagCloud from './tag_cloud';
+import tagCloudContainer from './tag_cloud_container.html';
```

Comment:
What made you choose to manually specify the container's html that is then modified using DOM methods/attributes as opposed to using React? This seems like it'd be a lot easier if we used React, and then we wouldn't have to manually ensure that we weren't opening ourselves up to XSS by ensure we're using the correct attributes on the elements themselves, and it'd be done automatically for us.

Reply:
Couple reasons imho:

What would React offer beyond adding another abstraction? What's the attack vector here that React would solve? The template is static content.

The async nature of the visualization is wrapped cleanly in the `async .render()` call instead of wrapping this in a React component and then having to work around it.

We recommend that community users use the `createBaseVisualization` since it is the most straightforward one (and the most simple interface to maintain). I think we should eat our own dogfood in a couple of places,  here as good a place as any.

Reply:
i think we should change that suggestion to use `createBaseVisualization` with kibana 6.3 and suggest to use `createReactVisualization`. and limit usage of `createBaseVisualization` only for when visualization is based on some other rendering library (like d3 or leaflet).

Reply:
^ very much disagree :)

Reply:
> What would React offer beyond adding another abstraction? What's the attack vector here that React would solve? The template is static content.

We're using `node.textContent` appropriately below when setting the label, but if we were to set `innerHtml` we open ourselves up to a potential XSS vulnerability (based on where the data comes from).

Additionally, we're dynamically setting styling throughout based on parameters when React provides a clear way to do so.

Reply:
Having to know about all of the selectors within the template is a disadvantage as well.

Reply:
`dangerouslySetInnerHtml` is just as dangerous :)

i think with tagcloud (if going react way) we would need to rewrite whole tagcloud library. If the goal is getting rid of angular i think this current approach is good enough.

Reply:
I think we should not at all impose a rendering-technology on our community developers. This almost as a matter of principle. We are not creating flexible interfaces when users need to follow the implementation-detail of the day. It used to be angular in Kibana, now it is React. Those are decisions that make sense for us (enforcing consistent tooling, enforcing consistent L&F ...), but have much less relevance for outside users.

Also, we need to really determine what the added value React brings as a wrapper for a Visualization-plugin. IMO this is very little for 80% of the visualizations. A visualization, almost by definition, is going to be not something textual or vanilla DOM-like and will rely on some other visualization library, that will introduce SVG, canvas, GL, clever use of HTML, ... What is React offering beyond a templating engine for that? Visualizations also should be a lot more asynchronous and stateful, so we can make efficient use of the DOM and animations. React doesn't really help with that, because those implementation details of how to do that efficiently will depend very much from visualization to visualization.

Reply:
btw: not saying we shouldn't rewrite this in react, just justificating my LGTM :)

Reply:
> dangerouslySetInnerHtml is just as dangerous :)

Agreed, but we wouldn't need that here. We could grab the `ref` and pass it to the TagCloud and let it do it's thing. It doesn't mean that the TagCloud itself couldn't do some nasty stuff, but at least the code that we currently "own" would get the default security mechanisms.

> Also, we need to really determine what the added value React brings as a wrapper for a Visualization-plugin.

As soon as we end up straying from standard HTML elements, I agree with you. However, in this situation a templating language for HTML is exactly what we need.


I'm all for us getting rid of AngularJS, but I'm not following why we're doing all this manually in this situation. Using React here would be less code, more consistent with the current direction and more secure by default.

Reply:
My comment regarding the discussion can be found in the general PR comments :-)